jobqueue_features/tests/test_decorators.py

```diff
@@ -1,41 +1,316 @@
+        def f():
+            pass
+
+        self._basic_tests(cluster_type=cluster_type, id_=_id)
+        controller.delete_cluster(_id)
+
+    def test_task_by_cluster_and_cluster_id(self):
+        _id = "test1"
+        cluster_type = CustomSLURMCluster
+        original_cluster = cluster_type(name=_id)
+
+        with self.assertRaises(ClusterException) as ctx:
+            @on_cluster(cluster=original_cluster)
+            @task(cluster=original_cluster, cluster_id="bad name")
+            def f():
+                pass
+        self.assertEqual("Cluster 'name' and cluster_id are different.",
+                         ctx.exception.message)
+
+        @on_cluster(cluster=original_cluster)
+        @task(cluster=original_cluster, cluster_id=_id)
+        def f():
+            pass
+
+        self._basic_tests(cluster_type=cluster_type, id_=original_cluster.name)
+        controller.delete_cluster(original_cluster.name)
+
+    def test_task_by_cluster_and_cluster_id_local(self):
+        _id = "test1"
+        cluster_type = LocalCluster
+        original_cluster = cluster_type()
+
+        @on_cluster(cluster=original_cluster, cluster_id=_id)
+        @task(cluster=original_cluster, cluster_id="_id")
+        def f():
+            pass
+
+        self._basic_tests(cluster_type=cluster_type, id_=_id)
+        controller.delete_cluster(id_=_id)
+
+    def _basic_tests(self, cluster_type, id_=None):
+        cluster, client = controller.get_cluster(id_)
+        self.assertIsInstance(cluster, cluster_type)
+        self.assertIsInstance(client, Client)
+        self.assertEqual(controller.default_cluster.__class__, cluster_type.__class__)
+
+
+class TestTaskExecutionLocalCluster(TestCase):
+    def setUp(self):
+        self.cluster_type = LocalCluster
+        set_default_cluster(self.cluster_type)
 
     def tearDown(self):
-        clusters_controller_singleton._close()
+        controller._close()
+
+    def test_init_blank(self):
+        @on_cluster()
+        def f():
+            pass
 
-    def test_init(self):
-        cluster, client = clusters_controller_singleton.get_cluster()
+        cluster, client = controller.get_cluster()
         self.assertIsInstance(cluster, self.cluster_type)
         self.assertIsInstance(client, Client)
 
     def test_call(self):
-        r = self.function()
+        @on_cluster()
+        @task()
+        def f():
+            return "test"
+        r = f()
         self.assertIsInstance(r, Future)
         self.assertEqual(r.result(), "test")
 
     def test_bad_task(self):
-        with self.assertRaises(ClusterException):
-
-            @task(cluster_id="bad_ID")
+        _id = "bad_ID"
+        with self.assertRaises(ClusterException) as ctx:
+            @task(cluster_id=_id)
             def bad_function():
                 return "test"
 
             bad_function()
+        self.assertEqual('No cluster "{}" set!'.format(_id),
+                         ctx.exception.message)
```

Comment:
From https://stackoverflow.com/a/43453668 you should probably use `str(ctx.exception)`

Reply:
Done